crowbar_framework/app/models/service_object.rb

```diff
@@ -682,9 +682,43 @@ def validate_proposal proposal
   #
   def validate_proposal_after_save proposal
     validate_proposal_constraints proposal
+    validate_postponed_nodes proposal
     handle_validation_errors
   end
 
+  #
+  # Make sure that when user wants to apply changes affecting the nodes that are postponed
+  # (= not upgraded while rest of the cloud is already upgraded)
+  # the nodes can be skipped.
+  #
+  def validate_postponed_nodes(proposal)
+    return unless upgrade_postponed?
+
+    skip_unready_nodes_enabled = Rails.application.config.experimental.fetch(
+      "skip_unready_nodes", {}
+    ).fetch("enabled", false)
+
+    skip_unready_nodes_roles = Rails.application.config.experimental.fetch(
+      "skip_unready_nodes", {}
+    ).fetch("roles", [])
+
+    postponed_nodes = []
+    proposal["deployment"][@bc_name]["elements"].each do |role, nodes|
+      next if skip_unready_nodes_enabled && skip_unready_nodes_roles.include?(role)
+      nodes.each do |n|
+        next if is_cluster? n
+        node = NodeObject.find_by_name(n)
+        postponed_nodes << n unless node.upgraded?
+      end
+    end
+    return if postponed_nodes.empty?
+
+    validation_error "The upgrade of some nodes has been postponed: " +
```

Comment:
won't it generate duplicates if there are multiple roles in one barclamp?

Reply:
I don't have any example at hand... but it doesn't mean that it can't happen :)

Reply:
I think it can't happen with our barclamps

Reply:
@skazi0 anyway, i've added `uniq` now.